File: mcp_servers/shared/validation.py
"""
Shared validation utilities for Klavis MCP servers.

Provides consistent environment variable validation and error handling
across all Python-based MCP servers.
"""

import os
import logging
from typing import Dict, List, Optional, Any
from dataclasses import dataclass
from enum import Enum


class ValidationLevel(Enum):
    """Validation level for environment variables."""
    REQUIRED = "required"
    OPTIONAL = "optional"
    CONDITIONAL = "conditional"


@dataclass
class EnvVarConfig:
    """Configuration for an environment variable."""
    name: str
    description: str
    validation_level: ValidationLevel
    default_value: Optional[str] = None
    setup_url: Optional[str] = None
    required_permissions: Optional[List[str]] = None
    depends_on: Optional[str] = None  # For conditional variables


class CredentialValidator:
    """Utility class for validating environment variables and handling auth errors."""
    
    def __init__(self, service_name: str, logger: Optional[logging.Logger] = None):
        self.service_name = service_name
        self.logger = logger or logging.getLogger(__name__)
    
    def validate_env_vars(self, env_vars: Dict[str, EnvVarConfig]) -> Dict[str, str]:
        """
        Validate environment variables according to their configuration.
        
        Args:
            env_vars: Dictionary mapping variable names to their configuration
            
        Returns:
            Dictionary of validated environment variables
            
        Raises:
            ValueError: If required environment variables are missing or invalid
        """
        validated_vars = {}
        missing_required = []
        missing_conditional = []
        
        for var_name, config in env_vars.items():
            value = os.getenv(var_name, config.default_value)
            
            if config.validation_level == ValidationLevel.REQUIRED:
                if not value or value.strip() == "":
                    missing_required.append(config)
                else:
                    validated_vars[var_name] = value
                    
            elif config.validation_level == ValidationLevel.CONDITIONAL:
                # Check if the conditional dependency is met
                if config.depends_on and os.getenv(config.depends_on):
                    if not value or value.strip() == "":
                        missing_conditional.append(config)
                    else:
                        validated_vars[var_name] = value
                elif value:
                    validated_vars[var_name] = value
                    
            else:  # OPTIONAL
                if value:
                    validated_vars[var_name] = value
        
        # Generate helpful error messages for missing variables
        if missing_required or missing_conditional:
            error_message = self._generate_error_message(missing_required, missing_conditional)
            raise ValueError(error_message)
        
        return validated_vars
    
    def _generate_error_message(self, missing_required: List[EnvVarConfig], 
                               missing_conditional: List[EnvVarConfig]) -> str:
        """Generate a helpful error message for missing environment variables."""
        lines = [f"❌ {self.service_name} MCP Server Configuration Error\n"]
        
        if missing_required:
            lines.append("Missing required environment variables:")
            for config in missing_required:
                lines.append(f"  • {config.name}: {config.description}")
                if config.setup_url:
                    lines.append(f"    Setup: {config.setup_url}")
                if config.required_permissions:
                    lines.append(f"    Required permissions: {', '.join(config.required_permissions)}")
                lines.append("")
        
        if missing_conditional:
            lines.append("Missing conditional environment variables:")
            for config in missing_conditional:
                lines.append(f"  • {config.name}: {config.description}")
                if config.depends_on:
                    lines.append(f"    Required when {config.depends_on} is set")
                lines.append("")
        
        lines.extend([
            "To fix this:",
            "1. Set the environment variables in your shell:",
            f"   export {missing_required[0].name if missing_required else missing_conditional[0].name}=your_token_here",
            "",
            "2. Or add to a .env file:",
            f"   {missing_required[0].name if missing_required else missing_conditional[0].name}=your_token_here",
            "",
            f"3. Restart the {self.service_name} MCP server",
        ])
        
        return "\n".join(lines)
    
    @staticmethod
    def create_auth_error_response(service: str, error_details: str, 
                                 status_code: int = 401) -> Dict[str, Any]:
        """
        Create a consistent error response for authentication failures.
        
        Args:
            service: Name of the service (e.g., "GitHub", "Slack")
            error_details: Specific error details
            status_code: HTTP status code (401 for auth, 403 for permissions)
            
        Returns:
            Standardized error response dictionary
        """
        return {
            "success": False,
            "error": f"{service} Authentication Error",
            "message": error_details,
            "status_code": status_code,
            "type": "authentication_error" if status_code == 401 else "authorization_error",
            "troubleshooting": {
                "check_token": f"Verify your {service} access token is valid",
                "check_permissions": f"Ensure your {service} token has required permissions",
                "check_expiration": f"Check if your {service} token has expired"
            }
        }
    
    @staticmethod
    def create_config_error_response(service: str, missing_vars: List[str]) -> Dict[str, Any]:
        """
        Create a consistent error response for configuration issues.
        
        Args:
            service: Name of the service
            missing_vars: List of missing environment variables
            
        Returns:
            Standardized error response dictionary
        """
        return {
            "success": False,
            "error": f"{service} Configuration Error",
            "message": f"Missing required environment variables: {', '.join(missing_vars)}",
            "status_code": 500,
            "type": "configuration_error",
            "missing_variables": missing_vars
        }


def validate_startup_config(service_name: str, env_vars: Dict[str, EnvVarConfig], 
                          logger: Optional[logging.Logger] = None) -> Dict[str, str]:
    """
    Convenience function for validating environment variables during server startup.
    
    Args:
        service_name: Name of the service for error messages
        env_vars: Dictionary of environment variable configurations
        logger: Optional logger instance
        
    Returns:
        Dictionary of validated environment variables
        
    Raises:
        ValueError: If validation fails with helpful error message
    """
    validator = CredentialValidator(service_name, logger)
    return validator.validate_env_vars(env_vars)


# Common environment variable configurations for reuse
COMMON_ENV_VARS = {
    "PORT": EnvVarConfig(
        name="PORT",
        description="Port for the MCP server to listen on",
        validation_level=ValidationLevel.OPTIONAL,
        default_value="5000"
    ),
    "LOG_LEVEL": EnvVarConfig(
        name="LOG_LEVEL", 
        description="Logging level (DEBUG, INFO, WARNING, ERROR)",
        validation_level=ValidationLevel.OPTIONAL,
        default_value="INFO"
    )
}

File: mcp_servers/shared/validation.ts
/**
 * Shared validation utilities for Klavis MCP servers.
 * 
 * Provides consistent environment variable validation and error handling
 * across all TypeScript/Node.js-based MCP servers.
 */

export enum ValidationLevel {
    REQUIRED = 'required',
    OPTIONAL = 'optional',
    CONDITIONAL = 'conditional'
}

export interface EnvVarConfig {
    name: string;
    description: string;
    validationLevel: ValidationLevel;
    defaultValue?: string;
    setupUrl?: string;
    requiredPermissions?: string[];
    dependsOn?: string; // For conditional variables
}

export interface AuthErrorResponse {
    success: boolean;
    error: string;
    message: string;
    statusCode: number;
    type: 'authentication_error' | 'authorization_error';
    troubleshooting: {
        checkToken: string;
        checkPermissions: string;
        checkExpiration: string;
    };
}

export interface ConfigErrorResponse {
    success: boolean;
    error: string;
    message: string;
    statusCode: number;
    type: 'configuration_error';
    missingVariables: string[];
}

export class CredentialValidator {
    private serviceName: string;
    private logger: { error: (msg: string) => void; info: (msg: string) => void };

    constructor(serviceName: string, logger?: any) {
        this.serviceName = serviceName;
        this.logger = logger || console;
    }

    /**
     * Validate environment variables according to their configuration.
     */
    validateEnvVars(envVars: Record<string, EnvVarConfig>): Record<string, string> {
        const validatedVars: Record<string, string> = {};
        const missingRequired: EnvVarConfig[] = [];
        const missingConditional: EnvVarConfig[] = [];

        for (const [varName, config] of Object.entries(envVars)) {
            const value = process.env[varName] || config.defaultValue;

            if (config.validationLevel === ValidationLevel.REQUIRED) {
                if (!value || value.trim() === '') {
                    missingRequired.push(config);
                } else {
                    validatedVars[varName] = value;
                }
            } else if (config.validationLevel === ValidationLevel.CONDITIONAL) {
                // Check if the conditional dependency is met
                if (config.dependsOn && process.env[config.dependsOn]) {
                    if (!value || value.trim() === '') {
                        missingConditional.push(config);
                    } else {
                        validatedVars[varName] = value;
                    }
                } else if (value) {
                    validatedVars[varName] = value;
                }
            } else { // OPTIONAL
                if (value) {
                    validatedVars[varName] = value;
                }
            }
        }

        // Generate helpful error messages for missing variables
        if (missingRequired.length > 0 || missingConditional.length > 0) {
            const errorMessage = this.generateErrorMessage(missingRequired, missingConditional);
            throw new Error(errorMessage);
        }

        return validatedVars;
    }

    private generateErrorMessage(missingRequired: EnvVarConfig[], missingConditional: EnvVarConfig[]): string {
        const lines: string[] = [`❌ ${this.serviceName} MCP Server Configuration Error\n`];

        if (missingRequired.length > 0) {
            lines.push('Missing required environment variables:');
            for (const config of missingRequired) {
                lines.push(`  • ${config.name}: ${config.description}`);
                if (config.setupUrl) {
                    lines.push(`    Setup: ${config.setupUrl}`);
                }
                if (config.requiredPermissions) {
                    lines.push(`    Required permissions: ${config.requiredPermissions.join(', ')}`);
                }
                lines.push('');
            }
        }

        if (missingConditional.length > 0) {
            lines.push('Missing conditional environment variables:');
            for (const config of missingConditional) {
                lines.push(`  • ${config.name}: ${config.description}`);
                if (config.dependsOn) {
                    lines.push(`    Required when ${config.dependsOn} is set`);
                }
                lines.push('');
            }
        }

        const firstMissing = missingRequired[0] || missingConditional[0];
        lines.push(
            'To fix this:',
            '1. Set the environment variables in your shell:',
            `   export ${firstMissing.name}=your_token_here`,
            '',
            '2. Or add to a .env file:',
            `   ${firstMissing.name}=your_token_here`,
            '',
            `3. Restart the ${this.serviceName} MCP server`
        );

        return lines.join('\n');
    }

    static createAuthErrorResponse(service: string, errorDetails: string, statusCode: number = 401): AuthErrorResponse {
        return {
            success: false,
            error: `${service} Authentication Error`,
            message: errorDetails,
            statusCode,
            type: statusCode === 401 ? 'authentication_error' : 'authorization_error',
            troubleshooting: {
                checkToken: `Verify your ${service} access token is valid`,
                checkPermissions: `Ensure your ${service} token has required permissions`,
                checkExpiration: `Check if your ${service} token has expired`
            }
        };
    }

    static createConfigErrorResponse(service: string, missingVars: string[]): ConfigErrorResponse {
        return {
            success: false,
            error: `${service} Configuration Error`,
            message: `Missing required environment variables: ${missingVars.join(', ')}`,
            statusCode: 500,
            type: 'configuration_error',
            missingVariables: missingVars
        };
    }
}

/**
 * Convenience function for validating environment variables during server startup.
 */
export function validateStartupConfig(
    serviceName: string, 
    envVars: Record<string, EnvVarConfig>, 
    logger?: any
): Record<string, string> {
    const validator = new CredentialValidator(serviceName, logger);
    return validator.validateEnvVars(envVars);
}

/**
 * Validate authentication token from request headers or environment.
 */
export function validateAuthToken(
    tokenFromHeader: string | undefined,
    envVarName: string,
    serviceName: string
): string {
    const token = tokenFromHeader || process.env[envVarName];
    
    if (!token || token.trim() === '') {
        throw new Error(
            `${serviceName} authentication token is missing. ` +
            `Provide it via ${envVarName} environment variable or x-auth-token header.`
        );
    }
    
    return token;
}

// Common environment variable configurations for reuse
export const COMMON_ENV_VARS: Record<string, EnvVarConfig> = {
    PORT: {
        name: 'PORT',
        description: 'Port for the MCP server to listen on',
        validationLevel: ValidationLevel.OPTIONAL,
        defaultValue: '5000'
    },
    LOG_LEVEL: {
        name: 'LOG_LEVEL',
        description: 'Logging level (debug, info, warn, error)',
        validationLevel: ValidationLevel.OPTIONAL,
        defaultValue: 'info'
    }
};